Microsoft.Identity.Web/Client/TokenCacheProviders/Redis/MSALAppRedisTokenCacheProvider.cs

```diff
@@ -0,0 +1,137 @@
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+SOFTWARE.
+***********************************************************************************************/
+
+using Microsoft.AspNetCore.Authentication.AzureAD.UI;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Caching.Distributed;
+using Microsoft.Extensions.Options;
+using Microsoft.Identity.Client;
+using System;
+
+namespace Microsoft.Identity.Web.Client.TokenCacheProviders
+{
+    /// <summary>
+    /// An implementation of token cache for both Confidential and Public clients backed by Redis Cache.
+    /// When cached data is distributed, the data:
+    /// Is coherent(consistent) across requests to multiple servers.
+    /// Survives server restarts and app deployments.
+    /// Doesn't use local memory.
+    /// </summary>
+    /// <seealso cref="https://aka.ms/msal-net-token-cache-serialization"/>
+    public class MSALAppRedisTokenCacheProvider : IMSALAppTokenCacheProvider
+    {
+        /// <summary>
+        /// The application cache key
+        /// </summary>
+        internal string AppCacheId;
+
+        /// <summary>
+        /// The backing IDistributedCache instance
+        /// </summary>
+        internal IDistributedCache distributedCache;
+
+        private readonly MSALRedisTokenCacheOptions CacheOptions;
+
+        /// <summary>
+        /// The App's whose cache we are maintaining.
+        /// </summary>
+        private readonly string AppId;
+
+        public MSALAppRedisTokenCacheProvider(IDistributedCache cache,
+            MSALRedisTokenCacheOptions option,
+            IOptionsMonitor<AzureADOptions> azureAdOptionsAccessor)
+        {
+            if (option == null)
+            {
+                throw new NullReferenceException("RedisCacheOptions are not provided");
+            }
+            else
+            {
+                this.CacheOptions = option;
+            }
+
+            if (azureAdOptionsAccessor.CurrentValue == null && string.IsNullOrWhiteSpace(azureAdOptionsAccessor.CurrentValue.ClientId))
+            {
+                throw new ArgumentNullException(nameof(AzureADOptions), $"The app token cache needs {nameof(AzureADOptions)}, populated with clientId to initialize.");
+            }
+
+            this.AppId = azureAdOptionsAccessor.CurrentValue.ClientId;
+            this.distributedCache = cache;
+        }
+
+        /// <summary>Initializes this instance of TokenCacheProvider with essentials to initialize themselves.</summary>
+        /// <param name="tokenCache">The token cache instance of MSAL application</param>
+        /// <param name="httpcontext">The Httpcontext whose Session will be used for caching.This is required by some providers.</param>
+        public void Initialize(ITokenCache tokenCache, HttpContext httpcontext)
```

Comment:
missing validation for a public API input parameter:
`if (tokenCache == null) throw new ArgumentNullException(nameof(tokenCache));`